### PR TITLE
DEP: Deprecate collapsing shape-1 dtype fields to scalars.

### PR DESCRIPTION
**Pull request by anntzer**
See discussion at https://github.com/numpy/numpy/issues/13112.

Currently, a field specified as `[(name, dtype, 1)]` is interpreted as
a scalar field (i.e., the same as `[(name, dtype)]` or `[(name, dtype,
()]`). This now raises a DeprecationWarning; in a future version, it
will be interpreted as a shape-(1,) field, i.e. the same as `[(name,
dtype, (1,))]` (consistently with `[(name, dtype, n)]` with `n>1`, which
is already equivalent to `[(name, dtype, (n,)]`).

@eric-wieser: This currently fails test_recfunctions::TestRecFunctions::test_structured_to_unstructured (apparently an extra copy is now being made), but it's not clear to me why.

### CODE REVIEW

File: numpy/core/src/multiarray/descriptor.c
Comment (eric-wieser) [inline numpy/core/src/multiarray/descriptor.c]: This should be a future warning, I believe we have a macro for that too.

**Comment by eric-wieser** [comment]
> This currently fails test_recfunctions::TestRecFunctions::test_structured_to_unstructured (apparently an extra copy is now being made), but it's not clear to me why.

Congrats, you found a bug in `astype(copy=False)`:

```python
dt = np.dtype([('f', (int, 3))])
a = np.zeros(3, dt)
assert a.astype(a.dtype, copy=False).base is not None  # fails
```

For the time being, you should probably just replace `(c,)` with `(c,) if c != 1 else ()` in the failing code paths.

File: numpy/lib/recfunctions.py
Comment (eric-wieser) [inline numpy/lib/recfunctions.py]: To work around #13327 you need to special-case `c == 1` here

Comment (eric-wieser) [reply numpy/lib/recfunctions.py]: Nevermind - the actual problem is the test is bad - it should be checking `ret.base is a or ret is a` - or in a shorter form, `ret.view().base is a`

Comment (anntzer) [reply numpy/lib/recfunctions.py]: Switching to either form still fails for me locally.

**Comment by eric-wieser** [comment]
Looking more closely - this is a bug in `structured_to_unstructured`

**Comment by eric-wieser** [comment]
Rebasing on #13332 should fix things

**Comment by anntzer** [comment]
I'll wait for it to be merged first?  Please ping me when it is :)

**Comment by mattip** [comment]
#1332 was merged, this needs a rebase

**Comment by anntzer** [comment]
thanks, rebased.

**Comment by mattip** [comment]
Looks good to me. Do we need to pass this by the mailing list?

File: numpy/core/src/multiarray/descriptor.c
Comment (eric-wieser) [inline numpy/core/src/multiarray/descriptor.c]: Deprecation warnings need a comment giving the version and date they were introduced. Search the source for use of the deprecate macro and you should see and example

**Comment by charris** [comment]
> Do we need to pass this by the mailing list?

Probably worth doing that. The proposed changes make sense to me, but it never hurts to see if anyone else has thoughts on the matter.

File: numpy/core/tests/test_dtype.py
Comment (eric-wieser) [inline numpy/core/tests/test_dtype.py]: Does changing this to (1)i4 fix the warning? We should try to write our tests to not rely on deprecated behavior for anything other than testing the deprecation itself.

File: numpy/lib/tests/test_recfunctions.py
Comment (eric-wieser) [inline numpy/lib/tests/test_recfunctions.py]: Is this necessary?

File: numpy/core/tests/test_dtype.py
Comment (eric-wieser) [inline numpy/core/tests/test_dtype.py]: Let's just fix the test here too. It's not trying to test the dtype construction here.

**Review by eric-wieser — CHANGES_REQUESTED** [review]
Needs a test in test_deprecations.py too, I think

File: numpy/core/tests/test_dtype.py
Comment (anntzer) [reply numpy/core/tests/test_dtype.py]: No, right now `"(1)i4"` is treated as `1i4` (both return a scalar i4 type).  IOW when specifying dtypes as strings there's no way to generate a shape-(1,) field.

File: numpy/core/tests/test_dtype.py
Comment (anntzer) [reply numpy/core/tests/test_dtype.py]: what dtype equality do you want to check here?  this checks that (dtype) is equivalent to (dtype, 1) which is exactly what this PR modifies.  we can't test that (dtype, 1) is equivalent to (dtype, (1,)) given that, well, that's simply not true yet.

File: numpy/core/src/multiarray/descriptor.c
Comment (anntzer) [reply numpy/core/src/multiarray/descriptor.c]: done

File: numpy/lib/tests/test_recfunctions.py
Comment (anntzer) [reply numpy/lib/tests/test_recfunctions.py]: not anymore (since the rebase), reverted.

**Comment by anntzer** [comment]
added test in test_deprecations

File: numpy/core/tests/test_dtype.py
Comment (eric-wieser) [reply numpy/core/tests/test_dtype.py]: You're right, my comment is misplaced here. I think I'd keep this test but with a comment explaining what the behavior will be once the deprecation expires.

Comment (anntzer) [reply numpy/core/tests/test_dtype.py]: done

File: numpy/core/tests/test_deprecations.py
Comment (eric-wieser) [inline numpy/core/tests/test_deprecations.py]: Should use self.assert_deprecated which checks twice, once for logging and once for raising (since those are different c code paths).

Comment (anntzer) [reply numpy/core/tests/test_deprecations.py]: done

File: numpy/core/src/multiarray/descriptor.c
Comment (eric-wieser) [inline numpy/core/src/multiarray/descriptor.c]: Based on your test changes, it might be worth a remark on the message that this also applies to / is caused by format strings like "1f4". I don't think you need to go to the effort of working out whether the user used tuples or strings though.

File: numpy/core/tests/test_dtype.py
Comment (eric-wieser) [reply numpy/core/tests/test_dtype.py]: (1,)i4 is also no good?

File: numpy/core/tests/test_deprecations.py
Comment (eric-wieser) [inline numpy/core/tests/test_deprecations.py]: I assume we never made a FutureDeprecationTestCase...

File: numpy/core/src/multiarray/descriptor.c
Comment (eric-wieser) [reply numpy/core/src/multiarray/descriptor.c]: And in the release note.

File: numpy/core/tests/test_dtype.py
Comment (anntzer) [reply numpy/core/tests/test_dtype.py]: Ah, (1,)i4 works, edited.

File: numpy/core/src/multiarray/descriptor.c
Comment (anntzer) [reply numpy/core/src/multiarray/descriptor.c]: good point, edited.

File: numpy/core/tests/test_deprecations.py
Comment (anntzer) [reply numpy/core/tests/test_deprecations.py]: grepping doesn't give anything

**Comment by mhvk** [comment]
In response to the mailing list: very much in favour!

**Comment by mattip** [comment]
Looks good to me. Can this make 1.17?

File: numpy/ma/mrecords.py
Comment (eric-wieser) [inline numpy/ma/mrecords.py]: I think this needs to be `(len(_mask.dtype),)` else it will run into the deprecation warning here when `len(_mask.dtype) == 1`

Comment (anntzer) [reply numpy/ma/mrecords.py]: good catch, fixed

File: numpy/ma/mrecords.py
Comment (eric-wieser) [inline numpy/ma/mrecords.py]: This if condition should be `_mask.dtype.fields is None`, else it blows up on structured types with 0 fields. Arguably out of scope, but may as well fix this.

If you're looking for another deprecation for another PR, len(dtype) is looking like a bad idea on non structured dtypes...

Comment (eric-wieser) [reply numpy/ma/mrecords.py]: Although actually the function is crap anyway and should be using one of the recursive helpers in ma/core that ors together all the fields

Comment (anntzer) [reply numpy/ma/mrecords.py]: So should I update this PR, or just leave it as it is and let you do a followup?  (I would prefer the latter :))

Comment (eric-wieser) [reply numpy/ma/mrecords.py]: I suppose you could argue that this PR is about fixing things for length-1 subarrays, and you've succeeded even without fixing this. I think fine to put in as is, assuming we can sort out the ci failure by restarting

**Comment by mattip** [comment]
CI is passing. Any more comments?

**Comment by anntzer** [comment]
Not on my side.

**Comment by mattip** [comment]
Still needs a test in `test_deprecations`

**Comment by anntzer** [comment]
it's already there? https://github.com/numpy/numpy/pull/13326/files#diff-5d1780949b81e83e8bf2b7b0b00c45e1

**Review by eric-wieser — APPROVED** [review]

**Comment by eric-wieser** [comment]
Let's put this in - more than one core dev in favor of the deprecation, and at worst this will create too much noise. Thanks @anntzer. I look forward to never being bitten by this again!

**Pull request merged**
